tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphFunction.java

```diff
@@ -0,0 +1,301 @@
+          outputs.size(),
+          resolveToOutput(outputs),
+          (byte[]) null,
+          null,
+          description,
+          status
+      );
+
+      status.throwExceptionIfNotOK();
+
+      return new GraphFunction(handle.withDeallocator(), inputs.size(), outputs.size());
+    }
+  }
+
+
+  /**
+   * Create a {@link NamedGraphFunction} with name-based argument and output resolution. Includes all ops in {@code
+   * body}, {@code inputs} and {@code outputs} must be the inputs and outputs of {@code body}.
+   *
+   * @param body the graph to use as the function body
+   * @param name the name of the function
+   * @param appendHash whether to append the has of the signature to the name
+   * @param inputs the inputs to the function and their names
+   * @param outputs the outputs to the function and their names
+   * @param description a human-readable description of the function.  May be null.
+   */
+  public static NamedGraphFunction create(Graph body, String name, boolean appendHash, Map<String, Operand<?>> inputs,
+      Map<String, Operand<?>> outputs, String description) {
+    List<Operand<?>> inputList = new ArrayList<>();
+    List<String> inputNames = new ArrayList<>();
+    for (Entry<String, Operand<?>> input : inputs.entrySet()) {
+      inputNames.add(input.getKey());
+      inputList.add(input.getValue());
+    }
+
+    List<Operand<?>> outputList = new ArrayList<>();
+    List<String> outputNames = new ArrayList<>();
+    for (Entry<String, Operand<?>> output : outputs.entrySet()) {
+      outputNames.add(output.getKey());
+      outputList.add(output.getValue());
+    }
+
+    try (PointerScope scope = new PointerScope()) {
+      TF_Status status = TF_Status.newStatus();
+
+      //TODO can't get output names working, likely due to passing it wrong
+
+//      PointerPointer<CharPointer> outputNamesPointer = new PointerPointer<>(outputNames.size());
+//      outputNamesPointer.putString(outputNames.toArray(new String[0]));
+
+      try (Reference ref = body.ref()) {
+        TF_Function handle = TF_GraphToFunction(
+            ref.nativeHandle(),
+            name,
+            (byte) (appendHash ? 1 : 0),
+            -1,
+            null,
+            inputList.size(),
+            resolveToOutput(inputList),
+            outputList.size(),
+            resolveToOutput(outputList),
+            (byte[]) null,
+            null,
+            description,
+            status
+        );
+
+        status.throwExceptionIfNotOK();
+
+        return new NamedGraphFunction(handle.withDeallocator(), inputNames, outputNames);
+      }
+    }
+  }
+
+  /**
+   * Create a {@link NamedGraphFunction} from a {@link ConcreteFunction}.
+   *
+   * @param function the {@code ConcreteFunction} to create the function from
+   * @param appendHash whether to append a hash of the signature to the name
+   */
+  public static NamedGraphFunction create(ConcreteFunction function, boolean appendHash) {
```

Comment:
I'm not sure to understand why do we need both `NamedGraphFunction` and `ConcreteFunction`, they seem to achieve pretty much the same goal, which is to provided named inputs/outputs on an encapsulated function, right?

In the Python paradigm, a `Function` is a method that wraps an executable graph (or eager code) while a `ConcreteFunction` is the realization of this function for specific type of input arguments (since Python is non-typed, the function is realized (or "concretized") when its invoke for the first time with a given set of input types.

In Java, as everything is typed, we might not need to make this distinction. So why not just calling it simply `Function` and merge both classes together? The only difference I see is that one runs a main `Graph` (using a session) while the other runs a secondary graph (using a ops). But there might be a way to abstract this to the user.

Reply:
Pretty much, yeah, but `NamedGraphFunction` has to manage the `TF_Function` handle and doesn't require a signature.  `ConcreteFunction` also allows for Tensor execution.  As for function, see #205, this is generally modeling Python's `DefinedFunction`/`_EagerDefinedFunction`/defun, which is separate from the `Function` class.

I could look into merging them, but it would make it a bit harder to declare the `GraphFunction`.  Would you be in favor of getting rid of the positional `GraphFunction` altogether and merging `NamedGraphFunction` with `ConcreteFunction`?

Reply:
Yes, I'm for merging everything together if that can simplify the API for the user. Or we can continue to segregate the logic into multiple classes but only one should be public.

Reply:
When trying to make `GraphFunction` (now called `DefinedFunction`) package local, I ran into the issue where I'd like to have a `Ops.call(ConcreteFunction, Map<String, Operand<?>)` function, but the necessary methods (and `DefinedFunction`) won't be accessible there.  There's a few options, as far as I can see:

1. Only have a call method on `ConcreteFunction`, like `ConcreteFunction.call(Ops tf, Map<String, Operand<?>)`.
2. Have a `@Operator` Function class in the main package
3. Have a public call method on `ConcreteFunction` that takes `Scope` instead of `Ops`, call that from the operator class
4. Move `ConcreteFunction`, `DefinedFunction`, the Operator class, and any future function classes to a new package (`function`, presumably).

4 is my preference, thoughts?